mrmustard/lab/abstract/transformation.py

```diff
@@ -142,13 +146,10 @@ def transform_fock(self, state: State, dual: bool) -> State:
     def modes(self) -> Sequence[int]:
         """Returns the list of modes on which the transformation acts on."""
         if self._modes in (None, []):
-            X, Y, d = self.XYd
-            if d is not None:
-                self._modes = list(range(d.shape[-1] // 2))
-            elif X is not None:
-                self._modes = list(range(X.shape[-1] // 2))
-            elif Y is not None:
-                self._modes = list(range(Y.shape[-1] // 2))
+            for elem in self.XYd:
+                if elem is not None:
+                    self._modes = list(range(elem.shape[-1] // 2))
+                    break
```

Comment:
I think every instance of `Transformation` should assign the property `_modes` given the number of parameters or the user-given modes; In that sense, `_modes=None` would be forbidden. On the other hand, as we discussed, `Transformation`s shouldn't have to deal with modes at all and the circuit should take care of it.

For now I'd leave here a TODO comment:
```suggestion
            # TODO: transfomations should set `_modes` instead of leaving the property as `None`
            for elem in self.XYd:
                if elem is not None:
                    self._modes = list(range(elem.shape[-1] // 2))
                    break
```